File: scripts/update-playlist.py
#!/usr/bin/env python3

import json
from pathlib import Path

# need to run:
# python3 scripts/update-playlist.py

PROJECT_ROOT = Path(__file__).resolve().parents[1]
MUSIC_DIR = PROJECT_ROOT / "music"
PLAYLIST_PATH = MUSIC_DIR / "playlist.json"
AUDIO_EXTENSIONS = {
    ".aac",
    ".aiff",
    ".alac",
    ".flac",
    ".m4a",
    ".mp3",
    ".ogg",
    ".opus",
    ".wav",
    ".webm",
}


def is_audio_file(path: Path) -> bool:
    return path.is_file() and not path.name.startswith(".") and path.suffix.lower() in AUDIO_EXTENSIONS


def main() -> None:
    tracks = []
    if MUSIC_DIR.exists():
      tracks = [
          path.relative_to(MUSIC_DIR).as_posix()
          for path in MUSIC_DIR.rglob("*")
          if is_audio_file(path)
      ]

    tracks.sort(key=str.casefold)
    PLAYLIST_PATH.write_text(json.dumps(tracks, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")
    print(f"Updated {PLAYLIST_PATH.relative_to(PROJECT_ROOT)} with {len(tracks)} track(s).")


if __name__ == "__main__":
    main()

File: scripts/update-playlist.js
#!/usr/bin/env node

// need to run:
// node scripts/update-playlist.js

const fs = require("fs");
const path = require("path");

const projectRoot = path.resolve(__dirname, "..");
const musicDir = path.join(projectRoot, "music");
const playlistPath = path.join(musicDir, "playlist.json");
const audioExtensions = new Set([
  ".aac",
  ".aiff",
  ".alac",
  ".flac",
  ".m4a",
  ".mp3",
  ".ogg",
  ".opus",
  ".wav",
  ".webm",
]);

function walk(dir) {
  if (!fs.existsSync(dir)) return [];
  return fs.readdirSync(dir, { withFileTypes: true }).flatMap((entry) => {
    if (entry.name.startsWith(".")) return [];
    const fullPath = path.join(dir, entry.name);
    if (entry.isDirectory()) return walk(fullPath);
    if (!entry.isFile()) return [];
    if (!audioExtensions.has(path.extname(entry.name).toLowerCase())) return [];
    return fullPath;
  });
}

const tracks = walk(musicDir)
  .map((file) => path.relative(musicDir, file).split(path.sep).join("/"))
  .sort((first, second) => first.localeCompare(second, "zh-CN", { numeric: true }));

fs.writeFileSync(playlistPath, `${JSON.stringify(tracks, null, 2)}\n`, "utf8");

console.log(`Updated ${path.relative(projectRoot, playlistPath)} with ${tracks.length} track(s).`);
